tinygrad/runtime/ops_webgpu.py

```diff
@@ -2,62 +2,280 @@
   def _copyin(self, dest, src: memoryview):
     if src.nbytes % 4:
       padded_src = bytearray(round_up(src.nbytes, 4))
       padded_src[:src.nbytes] = src
-    self.dev.queue.write_buffer(dest, 0, padded_src if src.nbytes % 4 else src)
+    write_buffer(self.dev, dest, 0, padded_src if src.nbytes % 4 else src)
   def _copyout(self, dest: memoryview, src):
-    buffer_data = self.dev.queue.read_buffer(src, 0)
-    dest[:] = buffer_data[:dest.nbytes] if src._nbytes > dest.nbytes else buffer_data
+    buffer_data = read_buffer(self.dev, src)
+    src_len = webgpu.wgpuBufferGetSize(src)
+    dest[:] = buffer_data[:dest.nbytes] if src_len  > dest.nbytes else buffer_data
+  def _free(self, opaque, options):
+    webgpu.wgpuBufferDestroy(opaque)
 
 class WebGpuDevice(Compiled):
   def __init__(self, device:str):
-    adapter = wgpu.gpu.request_adapter_sync(power_preference="high-performance")
-    timestamp_supported = wgpu.FeatureName.timestamp_query in adapter.features
-    wgpu_device = adapter.request_device_sync(required_features=[wgpu.FeatureName.timestamp_query] if timestamp_supported else [])
-    super().__init__(device, WebGpuAllocator(wgpu_device), WGSLRenderer(), Compiler(),
-                     functools.partial(WebGPUProgram, (wgpu_device, timestamp_supported)))
+    # Requesting an adapter
+    adapter_result: List[Any] = []
+
+    def adapter_cb(status, adapter, msg, _): adapter_result[:] = status, adapter, from_wgpu_str(msg)
+
+    cb_info = create_cb_info(webgpu.WGPURequestAdapterCallbackInfo, webgpu.WGPURequestAdapterCallback, adapter_cb)
+    wgpu_wait(webgpu.wgpuInstanceRequestAdapterF(instance,
+      webgpu.WGPURequestAdapterOptions(powerPreference=webgpu.WGPUPowerPreference_HighPerformance), cb_info))
+
+    if adapter_result[0] != webgpu.WGPURequestAdapterStatus_Success:
+      raise RuntimeError(f"Error requesting adapter: [{webgpu.WGPURequestAdapterStatus__enumvalues[adapter_result[0]]}] {adapter_result[2]}")
+
+    # Get supported features
+    supported_features = webgpu.WGPUSupportedFeatures()
+    webgpu.wgpuAdapterGetFeatures(adapter_result[1], supported_features)
+    timestamp_supported = "WGPUFeatureName_TimestampQuery" in [webgpu.WGPUFeatureName__enumvalues[supported_features.features[i]]
+      for i in range(supported_features.featureCount)]
+
+    required_features = [webgpu.WGPUFeatureName_ShaderF16]
+    if timestamp_supported: required_features.append(webgpu.WGPUFeatureName_TimestampQuery)
+
+    assert adapter_result[1] is not None, "adapter should not be none"
+    device_desc = webgpu.WGPUDeviceDescriptor()
+
+    # Disable "timestamp_quantization" for nanosecond precision: https://developer.chrome.com/blog/new-in-webgpu-120
+    toggle_desc = webgpu.WGPUDawnTogglesDescriptor()
+    toggle_desc.chain.sType = webgpu.WGPUSType_DawnTogglesDescriptor
+    toggle_desc.enabledToggleCount = 1
+    unsafe_apis = ctypes.cast(ctypes.pointer(to_c_string("allow_unsafe_apis")), ctypes.POINTER(ctypes.c_char))
```

Comment:
Hmm, I feel like some of this stuff can be made more readable and shorter with helpers